### PR TITLE
Fixes #380

### LINKED ISSUE #380: Make SparkCluster and SparkStep compatible to EMR release label 4.x
Currently spark cluster in using bootstrap actions to install spark, which worked fine in AMI 3.x, but not in release label 4.x. It needs to be updated to be able to work with release label 4.4 through configuration and change SparkStep accordingly.

### PR DESCRIPTION
**Pull request by hoangelos**

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,7 @@
 All notable changes to this project will be documented in this file.
 This project adheres to [Semantic Versioning](http://semver.org/).
 
+
```
Comment (sethyates) [inline CHANGELOG.md]: Please add comments on your change to the changelog.

File: core/src/main/scala/com/krux/hyperion/activity/SparkActivity.scala [hunk 2]
```diff
@@ -13,19 +14,24 @@ import com.krux.hyperion.resource.{ Resource, SparkCluster }
...
   activityFields: ActivityFields[SparkCluster],
+  jobRunner: HString,
+  scriptRunner: HString,
   steps: Seq[SparkStep],
   inputs: Seq[S3DataNode],
   outputs: Seq[S3DataNode],
   preStepCommands: Seq[HString],
   postStepCommands: Seq[HString]
-) extends EmrActivity[SparkCluster] {
+)(implicit hc: HyperionContext) extends EmrActivity[SparkCluster] {
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkActivity.scala]: We've tried to avoid passing the implicit to the case class constructor.  The `apply` should be sufficient for this?

File: core/src/main/scala/com/krux/hyperion/activity/SparkActivity.scala [hunk 3]
```diff
@@ -59,9 +65,27 @@ case class SparkActivity private (
 
 object SparkActivity extends RunnableObject {
 
-  def apply(runsOn: Resource[SparkCluster]): SparkActivity = new SparkActivity(
+  def jobRunner(runsOn: Resource[SparkCluster])(implicit hc: HyperionContext): HString = {
+    if(runsOn.asManagedResource.exists(_.isReleaseLabel4xx)) {
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkActivity.scala]: Space between `if` and `(`.

File: core/src/main/scala/com/krux/hyperion/activity/SparkActivity.scala [hunk 4]
```diff
@@ -59,9 +65,27 @@ case class SparkActivity private (
...
+      "command-runner.jar"
+    } else {
+      "s3://elasticmapreduce/libs/script-runner/script-runner.jar"
+    }
+  }
+
+  def apply(runsOn: Resource[SparkCluster])(implicit hc: HyperionContext): SparkActivity = new SparkActivity(
     baseFields = BaseFields(PipelineObjectId(SparkActivity.getClass)),
     activityFields = ActivityFields(runsOn),
+    jobRunner = jobRunner(runsOn)(hc),
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkActivity.scala]: Explicitly passing `hc` shouldn't be required?

File: core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala [hunk 5]
```diff
@@ -11,12 +11,15 @@ case class SparkStep private (
   jarUri: HS3Uri,
   mainClass: Option[MainClass],
   args: Seq[HString],
-  scriptRunner: HString,
-  jobRunner: HString,
+  var scriptRunner: HString,
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: Why are these marked as `var` now?

File: core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala [hunk 6]
```diff
@@ -51,8 +54,8 @@ object SparkStep {
     jarUri = jarUri,
     mainClass = None,
     args = Seq.empty,
-    scriptRunner = "s3://elasticmapreduce/libs/script-runner/script-runner.jar",
-    jobRunner = s"${hc.scriptUri}run-spark-step.sh",
+    scriptRunner = "",
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: More preferrably these would be marked as `Option[HString]` and defaulted to `None` rather than empty string.

File: core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala [hunk 7]
```diff
@@ -11,12 +11,15 @@ case class SparkStep private (
...
-  scriptRunner: HString,
-  jobRunner: HString,
+  var scriptRunner: HString,
+  var jobRunner: HString,
   sparkOptions: Seq[HString],
   sparkConfig: Map[HString, HString]
 ) {
 
+
+  def withJobRunner(runner: HString) = copy(jobRunner = runner)
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: `copy(jobRunner = Option(runner))` assuming you changed the property to `jobRunner: Option[HString]` as per below.

File: core/src/main/scala/com/krux/hyperion/activity/SparkTaskActivity.scala [hunk 8]
```diff
@@ -29,7 +27,7 @@ case class SparkTaskActivity private (
   outputs: Seq[S3DataNode],
   sparkOptions: Seq[HString],
   sparkConfig: Map[HString, HString]
-) extends EmrTaskActivity[SparkCluster] {
+)(implicit hc: HyperionContext) extends EmrTaskActivity[SparkCluster] {
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkTaskActivity.scala]: See comment about regarding implicit on the case class constructor.

File: core/src/main/scala/com/krux/hyperion/activity/SparkTaskActivity.scala [hunk 9]
```diff
@@ -55,6 +53,22 @@ case class SparkTaskActivity private (
   def withFiles(files: HString*) = withSparkOption(files.flatMap(file => Seq("--files": HString, file)): _*)
   def withMaster(master: HString) = withSparkOption("--master", master)
 
+  def jobRunner: HString = {
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkTaskActivity.scala]: See comments above on `SparkActivity`.  Would be nice to DRY this up between these two if possible.

File: core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala [hunk 10]
```diff
@@ -20,15 +20,32 @@ case class SparkCluster private (
...
   def updateResourceFields(fields: ResourceFields) = copy(resourceFields = fields)
   def updateEmrClusterFields(fields: EmrClusterFields) = copy(emrClusterFields = fields)
 
   def withSparkVersion(sparkVersion: HString) = copy(sparkVersion = sparkVersion)
 
+  override def withReleaseLabel(label: HString): Self = {
+    super.withReleaseLabel(label)
+    if (label.toString.startsWith("emr-4.")) {
+      this.withApplications("Spark": HString)
+      this.defaultBootstrapActions = Seq.empty[HString]
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala]: Rather than marking all of these as `var` and making this no longer an immutable object, why don't you simply `this.copy(defaultApplications = x, defaultBootstrapActions = y, etc)`?

File: core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala [hunk 11]
```diff
@@ -20,15 +20,32 @@ case class SparkCluster private (
 
   val logger = LoggerFactory.getLogger(SparkCluster.getClass)
 
+  var isReleaseLabel4xx = false
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala]: And what happens when there is a release label `5.x`?

File: core/src/main/scala/com/krux/hyperion/activity/SparkActivity.scala [hunk 12]
```diff
@@ -25,7 +28,15 @@ case class SparkActivity private (
   def updateBaseFields(fields: BaseFields) = copy(baseFields = fields)
   def updateActivityFields(fields: ActivityFields[SparkCluster]) = copy(activityFields = fields)
 
-  def withSteps(step: SparkStep*) = copy(steps = steps ++ step)
+  def withSteps(step: SparkStep*) = {
+    val new_steps = step.map(item =>
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkActivity.scala]: Please use camel case so `newSteps` not `new_steps`.

File: core/src/main/scala/com/krux/hyperion/activity/SparkActivity.scala [hunk 13]
```diff
@@ -25,7 +28,15 @@ case class SparkActivity private (
...
+      item.copy(
+        scriptRunner = Some(scriptRunner),
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkActivity.scala]: We prefer `Option` as a constructor and `Some` used only for matching.

File: core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala [hunk 14]
```diff
@@ -34,7 +37,7 @@ case class SparkStep private (
   def withMaster(master: HString) = withSparkOption("--master", master)
 
   def serialize: String = Seq(
-    Seq(scriptRunner, jobRunner),
+    Seq(scriptRunner.getOrElse(""), jobRunner.getOrElse("")),
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: Could you not simply have:
```
Seq(
  scriptRunner,
  jobRunner,
  sparkOptions,
  ...
)
```

File: core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala [hunk 15]
```diff
@@ -47,12 +50,18 @@ case class SparkStep private (
...
 object SparkStep {
 
-  def apply(jarUri: HS3Uri)(implicit hc: HyperionContext): SparkStep = SparkStep(
+  def apply(jarUri: HS3Uri)(implicit hc: HyperionContext): SparkStep = apply(
+    jarUri = jarUri,
+    scriptRunner = None,
+    jobRunner = None
+  )
+
+  def apply(jarUri: HS3Uri, jobRunner: Option[HString], scriptRunner: Option[HString])(implicit hc: HyperionContext): SparkStep = SparkStep(
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: Any reason not to have `jobRunner: Option[HString] = None, scriptRunner: Option[HString] = None)` rather than this and a separate `apply` without these parameters?

File: core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala [hunk 16]
```diff
@@ -47,12 +50,18 @@ case class SparkStep private (
...
     jarUri = jarUri,
     mainClass = None,
     args = Seq.empty,
-    scriptRunner = "s3://elasticmapreduce/libs/script-runner/script-runner.jar",
-    jobRunner = s"${hc.scriptUri}run-spark-step.sh",
+    scriptRunner = None,
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: Why aren't you passing in the parameters here and the next line?

File: core/src/main/scala/com/krux/hyperion/activity/SparkTaskActivity.scala [hunk 17]
```diff
@@ -19,8 +19,8 @@ case class SparkTaskActivity private (
   baseFields: BaseFields,
   activityFields: ActivityFields[SparkCluster],
   emrTaskActivityFields: EmrTaskActivityFields,
-  scriptRunner: HString,
   jobRunner: HString,
+  scriptRunner: HString,
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkTaskActivity.scala]: Why was this moved down?

File: core/src/main/scala/com/krux/hyperion/common/SparkCommandRunner.scala [hunk 18]
```diff
@@ -0,0 +1,24 @@
+package com.krux.hyperion.common
+
+import com.krux.hyperion.HyperionContext
+import com.krux.hyperion.adt.HString
+import com.krux.hyperion.resource.{SparkCluster, Resource}
+
+trait SparkCommandRunner {
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/common/SparkCommandRunner.scala]: Thanks!

File: core/src/main/scala/com/krux/hyperion/common/SparkCommandRunner.scala [hunk 19]
```diff
@@ -0,0 +1,24 @@
...
+  }
+
+  def scriptRunner(runsOn: Resource[SparkCluster]): HString = {
+    if (runsOn.asManagedResource.exists(_.isReleaseLabel4xx)) {
+      "command-runner.jar"
+    } else {
+      "s3://elasticmapreduce/libs/script-runner/script-runner.jar"
+    }
+  }
+}
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/common/SparkCommandRunner.scala]: Newline please.

File: core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala [hunk 20]
```diff
@@ -26,9 +30,25 @@ case class SparkCluster private (
...
   def withSparkVersion(sparkVersion: HString) = copy(sparkVersion = sparkVersion)
 
+  override def withReleaseLabel(label: HString): Self = {
+    super.withReleaseLabel(label)
+    if ((label.toString.length >= 6) && (label.toString.substring(4, 5).toInt >= 4)) {
+      //Assuming emr-x.y.z, and return x as a number.
+      this.copy(
+        defaultApplications = Seq("Spark": HString),
+        defaultBootstrapActions = Seq.empty[HString],
+        isReleaseLabel4xx = true
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala]: I still contend this is a bad name for this variable.  Doesn't the mere presence of having a releaseLabel trigger the different behaviour?  So this would be redundant and you could just use `releaseLabel.nonEmpty`, right?  I also don't think you need to do the string checking above as you wouldn't be getting this called unless the callee were using a release label.

File: core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala (hunk 15 above)
Comment (hoangelos) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: The reason I added a separate was to allow people to override the default. Although it's not possible so it does make sense. Or else I could change it to ignore setting it unless it is None.

Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: Yes but you can do the same thing with a single `apply` like so:

```
def apply(jarUri: HS3Uri, jobRunner: Option[HString] = None, scriptRunner: Option[HString] = None)(implicit hc: HyperionContext): SparkStep = ...
```

File: core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala (hunk 20 above)
Comment (hoangelos) [inline core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala]: Technically it isn't just the lack of the exist an e but at least 4.x. Or greater right?

File: core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala (hunk 15 above)
Comment (hoangelos) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: Of course. But what do you think about SparkActivity not updating it if SoarkStep already is set?  That way users can overwrite with their own?

File: core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala (hunk 20 above)
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala]: Release labels only supported >= 4.x.  My reading of this is pre-release labels (aka pre 4.x), we did one thing, afterwards, we do it a different way.

File: core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala (hunk 15 above)
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: OK but this `apply` is about creating new `SparkStep`s.  The not-updating would be handled in `SparkActivity` itself.  I'll post a comment there on how it would appear.  I'm actually unsure if this is a real use-case though...

Comment (hoangelos) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: It is for us. ;)

File: core/src/main/scala/com/krux/hyperion/activity/SparkActivity.scala (hunk 13 above)
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkActivity.scala]: If you wanted to keep what was specified in the SparkStep and only apply these as a "default", you could use: `item.copy(scriptRunner = item.scriptRunner.orElse(Option(scriptRunner)), ...)`.

File: core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala (hunk 15 above)
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: Cool! :-)

File: core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala (hunk 14 above)
Comment (hoangelos) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: If I do that it outputs the Option instead of the value, which would affect the JSON wouldn't it?

File: core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala (hunk 20 above)
Comment (hoangelos) [inline core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala]: I can convert to using that.  I assumed this was something inherent in emr-4.x.x, but you are saying it's more backwards compatibility with hyperion library.

File: core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala [hunk 21]
```diff
@@ -13,7 +13,11 @@ case class SparkCluster private (
   baseFields: BaseFields,
   resourceFields: ResourceFields,
   emrClusterFields: EmrClusterFields,
-  sparkVersion: HString
+  sparkVersion: HString,
+  defaultApplications: Seq[HString],
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala]: defaultApplications, defaultBootstrapActions and hasReleaseLabel are not required.

File: core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala [hunk 22]
```diff
@@ -26,9 +30,24 @@ case class SparkCluster private (
 
   def withSparkVersion(sparkVersion: HString) = copy(sparkVersion = sparkVersion)
 
+  override def withReleaseLabel(label: HString): Self = {
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala]: Remove this.

File: core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala [hunk 23]
```diff
@@ -26,9 +30,24 @@ case class SparkCluster private (
...
+        defaultApplications = Seq("Spark": HString),
+        defaultBootstrapActions = Seq.empty[HString],
+          hasReleaseLabel = true
+      )
+    } else {
+      this
+    }
+  }
+
+  override def applications =
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala]:
```
override def applications = if (releaseLabel.nonEmpty)
  ("Spark": HString) +: super.applications
else
  super.applications
```

File: core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala [hunk 24]
```diff
@@ -26,9 +30,24 @@ case class SparkCluster private (
...
+    this.defaultApplications ++ super.applications
+
   override def standardBootstrapAction = 
-    (s"s3://support.elasticmapreduce/spark/install-spark,-v,${sparkVersion},-x": HString) +:
-    super.standardBootstrapAction
+    this.defaultBootstrapActions ++ super.standardBootstrapAction
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala]:
```
override def standardBootstrapAction = if (releaseLabel.nonEmpty)
  super.standardBootstrapAction
else
  (s"s3://support.elasticmapreduce/spark/install-spark,-v,${sparkVersion},-x": HString) +: super.standardBootstrapAction
```

File: core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala [hunk 25]
```diff
@@ -38,7 +57,10 @@ object SparkCluster {
     baseFields = BaseFields(PipelineObjectId(SparkCluster.getClass)),
     resourceFields = EmrCluster.defaultResourceFields(hc),
     emrClusterFields = EmrCluster.defaultEmrClusterFields(hc),
-    sparkVersion = hc.emrSparkVersion.get
+    sparkVersion = hc.emrSparkVersion.get,
+    defaultApplications = Seq.empty[HString],
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/resource/SparkCluster.scala]: defaultApplications, defaultBootstrapActions and hasReleaseLabel are not required.

File: core/src/main/scala/com/krux/hyperion/common/SparkCommandRunner.scala [hunk 26]
```diff
@@ -0,0 +1,24 @@
+package com.krux.hyperion.common
+
+import com.krux.hyperion.HyperionContext
+import com.krux.hyperion.adt.HString
+import com.krux.hyperion.resource.{SparkCluster, Resource}
+
+trait SparkCommandRunner {
+
+  def jobRunner(runsOn: Resource[SparkCluster])(implicit hc: HyperionContext): HString = {
+    if (runsOn.asManagedResource.exists(_.hasReleaseLabel)) {
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/common/SparkCommandRunner.scala]:
```
if (runsOn.asManagedResource.exists(_.releaseLabel.nonEmpty)) {
```

File: core/src/main/scala/com/krux/hyperion/common/SparkCommandRunner.scala [hunk 27]
```diff
@@ -0,0 +1,24 @@
...
+      "spark-submit"
+    } else {
+      s"${hc.scriptUri}run-spark-step.sh"
+    }
+  }
+
+  def scriptRunner(runsOn: Resource[SparkCluster]): HString = {
+    if (runsOn.asManagedResource.exists(_.hasReleaseLabel)) {
```
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/common/SparkCommandRunner.scala]:
```
if (runsOn.asManagedResource.exists(_.releaseLabel.nonEmpty)) {
```

File: core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala (hunk 14 above)
Comment (sethyates) [inline core/src/main/scala/com/krux/hyperion/activity/SparkStep.scala]: Got it.

**Comment by sethyates** [comment]
Hi @hoangelos, I've looked into this more and I think the initial approach won't cover all cases (such as for some reason setting `withAmiVersion` after using `withReleaseLabel`).  I've made some recommendations above that should actually simplify this and handle all these cases.  Let me know what you think.

**Comment by sethyates** [comment]
Replaced this PR with #427
